### PR TITLE
feat(deployment): add deployment table

### PR DESCRIPTION
**Pull request by claire1618**
When a repo has a lot of deployments historically, viewing them in the UI is cumbersome and slow.

The loading of deployments on the deployments tab in Vela takes a few more seconds to load than the other pages (schedules, secrets, audits...) even when there are no deployments for a repo. This is due to the way deployments are stored and retrieved.

The changes made in this pull request add a new database table so that we can store deployment information instead of having to go to the scm to retrieve them. This (plus the implementation in server) greatly improve the time it takes for the deployments page to load.

### CODE REVIEW

**Comment by wass3rw3rk** [comment]
release and other merges caused some conflicts, so you'll have to resolve that first

**Comment by wass3rw3rk** [comment]
also, we will want to add tests

**Review by KellyMerrick — CHANGES_REQUESTED** [review]

**Review by timhuynh94 — APPROVED** [review]

File: database/deployment.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	URL         sql.NullString     `sql:"url"`
+	User        sql.NullString     `sql:"user"`
+	Commit      sql.NullString     `sql:"commit"`
+	Ref         sql.NullString     `sql:"ref"`
+	Task        sql.NullString     `sql:"task"`
+	Target      sql.NullString     `sql:"target"`
+	Description sql.NullString     `sql:"description"`
+	Payload     raw.StringSliceMap `sql:"payload"`
+	Builds      pq.StringArray     `sql:"builds" gorm:"type:varchar(50)"`
+}
```
Comment (wass3rw3rk) [inline database/deployment.go:39]: suggestion: we might want to add Created as we do in other tables so we can more easily purge these

**Review by wass3rw3rk — CHANGES_REQUESTED** [review]

File: database/deployment.go (hunk 1 above)
Comment (claire1618) [reply database/deployment.go]: added!

File: database/build.go [hunk 2]
```diff
@@ -45,6 +45,7 @@ type Build struct {
 	Started       sql.NullInt64      `sql:"started"`
 	Finished      sql.NullInt64      `sql:"finished"`
 	Deploy        sql.NullString     `sql:"deploy"`
+	DeploymentID  sql.NullInt64      `sql:"deployment_id"`
```
Comment (ecrupper) [inline database/build.go:48]:
```suggestion
	DeployID  sql.NullInt64      `sql:"deployment_id"`
```
Thoughts on this for consistency with `Deploy` and `DeployPayload`?

File: database/build.go [hunk 3]
```diff
@@ -170,6 +171,11 @@ func (b *Build) Nullify() *Build {
 		b.Deploy.Valid = false
 	}
 
+	// check if the Deploy field should be false
```
Comment (ecrupper) [inline database/build.go:174]:
```suggestion
	// check if the DeployID field should be false
```

File: database/deployment.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	d.Ref = sql.NullString{String: sanitize(d.Ref.String), Valid: d.Ref.Valid}
+	d.Task = sql.NullString{String: sanitize(d.Task.String), Valid: d.Task.Valid}
+	d.Target = sql.NullString{String: sanitize(d.Target.String), Valid: d.Target.Valid}
+	d.Description = sql.NullString{String: sanitize(d.Description.String), Valid: d.Description.Valid}
+
+	return nil
+}
+
+// DeploymentFromLibrary converts the Deployment type
+// to a library Deployment type.
```
Comment (ecrupper) [inline database/deployment.go:159]:
```suggestion
// DeploymentFromLibrary converts the library Deployment type
// to a database Deployment type.
```

File: database/deployment_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	want.SetTarget("production")
+	want.SetDescription("Deployment request from Vela")
+	want.SetPayload(raw.StringSliceMap{"foo": "test1"})
+	want.SetCreatedAt(1)
+	want.SetCreatedBy("octocat")
+	want.SetBuilds(nil)
+
+	got := testDeployment().ToLibrary(nil)
+	if diff := cmp.Diff(got, want); diff != "" {
+		t.Errorf("(ToLibrary: -want +got):\n%s", diff)
```
Comment (ecrupper) [inline database/deployment_test.go:83]: The order of want and got have to match in the error message as they do in the function call

File: database/deployment_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	d.SetURL("https://github.com/github/octocat/deployments/1")
+	d.SetCommit("1234")
+	d.SetRef("refs/heads/main")
+	d.SetTask("deploy:vela")
+	d.SetTarget("production")
+	d.SetDescription("Deployment request from Vela")
+	d.SetPayload(raw.StringSliceMap{"foo": "test1"})
+	d.SetCreatedAt(1)
+	d.SetCreatedBy("octocat")
+	d.SetBuilds(nil)
```
Comment (ecrupper) [inline database/deployment_test.go:145]: I'd like to see some unit testing on when this field is actually populated. Just a simple build partial or two should work as long as the ID is set.

File: library/build.go [hunk 7]
```diff
@@ -158,6 +159,7 @@ func (b *Build) Environment(workspace, channel string) map[string]string {
 		envs["VELA_BUILD_TARGET"] = target
 		envs["VELA_DEPLOYMENT"] = target
 		envs["BUILD_TARGET"] = target
+		envs["VELA_DEPLOYMENT_NUMBER"] = ToString(b.GetDeploymentID())
```
Comment (ecrupper) [inline library/build.go:162]: One of the benefits of storing our own deployment record is we can create our own `Environment` function for the deployment type. Might be worth considering here. At bare minimum, this should be `VELA_DEPLOYMENT_ID`, which is arguably not worth including 😅

**Review by ecrupper — CHANGES_REQUESTED** [review]
A few random things I noticed + thoughts on giving `library.Deployment` its own `Environment()` method? I can imagine that could be useful to some users.

File: library/build.go (hunk 7 above)
Comment (wass3rw3rk) [inline library/build.go:162]:
```suggestion
		envs["VELA_DEPLOYMENT_ID"] = ToString(b.GetDeploymentID())
```

File: library/build_test.go [hunk 8]
```diff
@@ -226,6 +228,7 @@ func TestLibrary_Build_Environment(t *testing.T) {
 				"VELA_BUILD_TITLE":         "push received from https://github.com/github/octocat",
 				"VELA_BUILD_WORKSPACE":     "TODO",
 				"VELA_DEPLOYMENT":          "production",
+				"VELA_DEPLOYMENT_NUMBER":   "0",
```
Comment (wass3rw3rk) [inline library/build_test.go:231]:
```suggestion
				"VELA_DEPLOYMENT_ID":   "0",
```

File: library/build_test.go [hunk 9]
```diff
@@ -286,6 +289,7 @@ func TestLibrary_Build_Environment(t *testing.T) {
...
```
Comment (wass3rw3rk) [inline library/build_test.go:292]:
```suggestion
				"VELA_DEPLOYMENT_ID":   "0",
```

**Review by wass3rw3rk — CHANGES_REQUESTED** [review]
one suggestion.. got a little confused with ID and Number 😵‍💫 let me know what you think - looks like Easton mentioned this also :disappear:

File: library/build.go (hunk 7 above)
Comment (wass3rw3rk) [reply library/build.go:162]: that's fair - i like that

**Review by wass3rw3rk — APPROVED** [review]

**Review by plyr4 — APPROVED** [review]
looks great.

File: library/deployment.go [hunk 10]
```diff
@@ -13,16 +13,18 @@ import (
 // swagger:model Deployment
 type Deployment struct {
 	ID          *int64              `json:"id,omitempty"`
+	Number      *int64              `json:"number,omitempty"`
 	RepoID      *int64              `json:"repo_id,omitempty"`
 	URL         *string             `json:"url,omitempty"`
-	User        *string             `json:"user,omitempty"`
```
Comment (plyr4) [inline library/deployment.go:18]: just curious, why'd we remove User?

**Review by ecrupper — APPROVED** [review]

File: library/deployment.go (hunk 10 above)
Comment (claire1618) [reply library/deployment.go:18]: because I added a CreatedBy field instead

**Review by KellyMerrick — APPROVED** [review]
lgtm

**Pull request merged**